File: rust/collatz.rs
// @DrParanoya on 12/04/2025
// Why does it take 5 lines to read an integer from stdin? Rust is even worse than C++

use std::time::Instant;
use std::io::stdin;
use std::io::{self, Write};

fn main() {
    print!("Please enter a positive integer to run the collatz conjecture on: ");
    io::stdout().flush().expect("Could not flush stdout");

    let mut input_line = String::new();
    stdin()
        .read_line(&mut input_line)
        .expect("Failed to read line");
    let mut number: u64 = input_line.trim().parse().expect("Invalid input!");

    let mut peak: u64 = number;
    let mut steps: u64 = 0;

    let start = Instant::now();
    while number != 1 {
        if number % 2 == 0 {
            number /= 2;
        } else {
            number = number * 3 + 1;
        }
        if number > peak {
            peak = number;
        }
        steps += 1;
    }
    println!("Steps to reach 1: {}\nPeak number reached: {}\nExecution time: {:.5?}", steps, peak, start.elapsed());
}

File: fortran/collatz.f90
! Written with <3 by @DrParanoya on 15/07/2026 (DD/MM/YYYY)
! I really like Fortran, it feels similar to COBOL, but way more modern and way better to use.

program read_values
  implicit none
  integer :: number, steps, peak
  real :: start, end

  print *, 'Please enter a positive integer to run the collatz conjecture on: '
  read(*,*) number
  if (number <= 0) then
    print *, 'Invalid input!'
	stop
  end if

  steps = 0
  peak = number

  call cpu_time(start)

  do while (number /= 1)
	if (mod(number, 2) == 0) then
      number = number / 2
	else
      number = number * 3 + 1
	end if

	if (number > peak) then
	  peak = number
	end if
      
	steps = steps + 1
  end do

  call cpu_time(end)

  print *, 'Steps to reach 1: ', steps 
  print *,'Peak number reached: ', peak
  print *, 'Time taken: ', end - start, ' seconds'

end program read_values
